runelite-client/src/main/java/net/runelite/client/plugins/puzzlesolver/PuzzleSolverOverlay.java

```diff
@@ -216,10 +218,36 @@ else if (config.displayRemainingMoves())
 						{
 							if (config.drawDots())
 							{
-								graphics.setColor(Color.YELLOW);
+								int moves = config.movesToShow();
+								if (moves > 4)
+								{
+									// positions capacity chosen so set does not reallocate when added to
+									Set<Integer> positions = new HashSet<>(4 * moves / 3);
+									for (int i = 1; i <= moves; i++)
+									{
+										int j = solver.getPosition() + i;
+
+										if (j >= solver.getStepCount())
+										{
+											moves = i - 1;
+											break;
+										}
+
+										PuzzleState futureMove = solver.getStep(j);
+
+										if (futureMove == null || !config.drawOverlaps() && positions.contains(futureMove.getEmptyPiece()))
+										{
+											moves = i - 1;
+											break;
+										}
+										positions.add(futureMove.getEmptyPiece());
+									}
+								}
+								moves = Math.min(4, moves); // Scale the dot sizing/color interpolation by at least 4
```

Comment:
I assume this moves variable is so circle #0 is `dotColor` and the #movesToShow circle is definitely `gradientColour`? Regardless if you have many steps left or only two. Correct me if I'm wrong.

Well, this adds extra complexity. Would it be ok to always interpolate between dotColour and gradientColour in `movesToShow` steps instead? That way you could take the original for loop and just do the positions.contains early termination without having to calculate the no. of moves ahead of time.